contracts/SDUtilityPool.sol

```diff
@@ -0,0 +1,906 @@
+        uint256 sdRequested = (exchangeRate * _cTokenAmount) / DECIMAL;
+        _requestId = _requestWithdraw(sdRequested, _cTokenAmount);
+    }
+
+    /**
+     * @notice auxiliary method to put a withdrawal request, takes SD amount in input
+     * @param _sdAmount amount of SD to withdraw
+     * @return _requestId generated request ID for withdrawal
+     */
+    function requestWithdrawWithSDAmount(uint256 _sdAmount)
+        external
+        override
+        whenNotPaused
+        returns (uint256 _requestId)
+    {
+        accrueFee();
+        uint256 exchangeRate = _exchangeRateStoredInternal();
+        uint256 cTokenToReduce = (_sdAmount * DECIMAL) / exchangeRate;
+        if (cTokenToReduce > delegatorCTokenBalance[msg.sender] - delegatorWithdrawRequestedCTokenCount[msg.sender]) {
+            revert InvalidAmountOfWithdraw();
+        }
+        delegatorWithdrawRequestedCTokenCount[msg.sender] += cTokenToReduce;
+        _requestId = _requestWithdraw(_sdAmount, cTokenToReduce);
+    }
+
+    /**
+     * @notice finalize delegator's withdraw requests
+     */
+    function finalizeDelegatorWithdrawalRequest() external override whenNotPaused {
+        accrueFee();
+        uint256 exchangeRate = _exchangeRateStoredInternal();
+        uint256 maxRequestIdToFinalize = Math.min(nextRequestId, nextRequestIdToFinalize + finalizationBatchLimit) - 1;
+        uint256 requestId;
+        uint256 sdToReserveToFinalizeRequests;
+        for (requestId = nextRequestIdToFinalize; requestId <= maxRequestIdToFinalize; ) {
+            DelegatorWithdrawInfo memory delegatorWithdrawInfo = delegatorWithdrawRequests[requestId];
+            uint256 requiredSD = delegatorWithdrawInfo.sdExpected;
+            uint256 amountOfcToken = delegatorWithdrawInfo.amountOfCToken;
+            uint256 minSDRequiredToFinalizeRequest = Math.min(requiredSD, (amountOfcToken * exchangeRate) / DECIMAL);
+            if (
+                (sdToReserveToFinalizeRequests + minSDRequiredToFinalizeRequest > getPoolAvailableSDBalance()) ||
+                (delegatorWithdrawInfo.requestBlock + minBlockDelayToFinalizeRequest > block.number)
+            ) {
+                break;
+            }
+            delegatorWithdrawRequests[requestId].sdFinalized = minSDRequiredToFinalizeRequest;
+            sdRequestedForWithdraw -= requiredSD;
+            sdToReserveToFinalizeRequests += minSDRequiredToFinalizeRequest;
+            delegatorCTokenBalance[delegatorWithdrawInfo.owner] -= amountOfcToken;
+            delegatorWithdrawRequestedCTokenCount[delegatorWithdrawInfo.owner] -= amountOfcToken;
+            cTokenTotalSupply -= amountOfcToken;
+            unchecked {
+                ++requestId;
+            }
+        }
+        // at here, upto (requestId-1) is finalized
+        if (requestId > nextRequestIdToFinalize) {
+            nextRequestIdToFinalize = requestId;
+            sdReservedForClaim += sdToReserveToFinalizeRequests;
```

Comment:
You should move this stmt into the for loop. Will reduce code complexity.

Line 174 and 181 get simplified. Plus get to remove this line.
Also don't see a harm in removing the `if` and assign ` nextRequestIdToFinalize = requestId;` directly.